lib/pbench/server/api/resources/query_apis/datasets_publish.py

```diff
@@ -0,0 +1,207 @@
+from typing import Any, AnyStr, Dict
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources.query_apis import (
+    ElasticBase,
+    Schema,
+    Parameter,
+    ParamType,
+)
+from pbench.server.database.models.tracker import Dataset, Metadata
+
+
+class DatasetsPublish(ElasticBase):
+    """
+    Change the "access" authorization of a Pbench dataset
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(
+                Parameter("controller", ParamType.STRING, required=True),
+                Parameter("name", ParamType.STRING, required=True),
+                Parameter("user", ParamType.USER, required=True, nullable=False),
+                Parameter("access", ParamType.ACCESS, required=True),
+            ),
+        )
+
+    def preprocess(self, client_json: Dict[AnyStr, Any]) -> Dict[AnyStr, Any]:
+        """
+        Query the Dataset associated with this name, and determine whether the
+        "access" is already configured to the value specified by the client. If
+        so, there's no need for an Elasticsearch query.
+
+        Otherwise, record the dataset so that the postprocess operation can
+        mark it as published.
+
+        Args:
+            client_json: Request JSON payload
+
+        Raises:
+            Any errors in the postprocess method shall be reported by
+            exceptions which will be logged and will terminate the operation.
+
+        Returns:
+            dataset object if the operation should continue, or None
+        """
+        dataset = Dataset.attach(
+            controller=client_json["controller"], name=client_json["name"]
+        )
+
+        # if the Dataset already has the requested access, succeed without
+        # doing any additional work
+        if dataset.access == client_json["access"]:
+            self.logger.info(
+                "No action necessary: {} is already access={}", dataset, dataset.access
+            )
+            return None
+
+        # The dataset exists, and doesn't already have the desired access, so
+        # we can go ahead
+        return dataset
+
+    def assemble(
+        self, json_data: Dict[AnyStr, Any], context: Dataset
+    ) -> Dict[AnyStr, Any]:
+        """
+        Construct an Elastsearch query to update the "access" field of the
+        "authorization" subdocument based on the client data.
+
+        {
+            "name": "dataset name",
+            "controller": "controller name",
+            "user": "username",
+            "access": "desired access"
+        }
+
+        json_data: JSON dictionary of type-normalized parameters
+            user: specifies the owner of the data to be modified; it need not
+                necessarily be the user represented by the session token
+                header, assuming the authenticated user can access "user"s
+                data.
```

Comment:
Interesting point. When I started this, I thought in terms of doing a query to find all the documents; the query complication became overwhelming, and instead I built the map, so we don't care.

However ... I ought to have gone back and checked the ownership and access of the `Dataset` against the authorization token and failed the operation in `preprocess` if the caller can't publish this dataset.

So that's a serious bug that I completely overlooked. So instead of removing "user", I'm going to have to add the code to use it!